tests/pass/string.glu

```diff
@@ -1,58 +1,54 @@
 let find_tests =
-    assert_oieq (string.find "abcd1234" "ab") (Some 0)
-        *> assert_oieq (string.find "abcd1234" "b") (Some 1)
-        *> assert_oieq (string.find "abcd1234" "4") (Some 7)
-        *> assert_oieq (string.find "abcd1234" "xyz") None
-        *> assert_oieq (string.rfind "abcdabcd" "b") (Some 5)
-        *> assert_oieq (string.rfind "abcdabcd" "d") (Some 7)
-        *> assert_oieq (string.rfind "abcd1234" "xyz") None
+    assert_eq (string.find "abcd1234" "ab") (Some 0)
+        *> assert_eq (string.find "abcd1234" "b") (Some 1)
+        *> assert_eq (string.find "abcd1234" "4") (Some 7)
+        *> assert_eq (string.find "abcd1234" "xyz") None
+        *> assert_eq (string.rfind "abcdabcd" "b") (Some 5)
+        *> assert_eq (string.rfind "abcdabcd" "d") (Some 7)
+        *> assert_eq (string.rfind "abcd1234" "xyz") None
 
 let starts_ends_tests =
-    assert_beq (string.starts_with "abcd1234" "ab") True
-        *> assert_beq (string.starts_with "abcd1234" "b") False
-        *> assert_beq (string.ends_with "abcd1234" "1234") True
-        *> assert_beq (string.ends_with "abcd1234" "4") True
-        *> assert_beq (string.ends_with "abcd1234" "ab") False
+    assert_eq (string.starts_with "abcd1234" "ab") True
+        *> assert_eq (string.starts_with "abcd1234" "b") False
+        *> assert_eq (string.ends_with "abcd1234" "1234") True
+        *> assert_eq (string.ends_with "abcd1234" "4") True
+        *> assert_eq (string.ends_with "abcd1234" "ab") False
 
 let trim_tests =
-    assert_seq (string.trim "ab") "ab" *> assert_seq (string.trim " ab ") "ab"
-        *> assert_seq (string.trim "ab \t") "ab"
-        *> assert_seq (string.trim "\t ab") "ab"
-        *> assert_seq (string.trim_left " ab ") "ab "
-        *> assert_seq (string.trim_right " ab ") " ab"
+    assert_eq (string.trim "ab") "ab" *> assert_eq (string.trim " ab ") "ab"
+        *> assert_eq (string.trim "ab \t") "ab"
+        *> assert_eq (string.trim "\t ab") "ab"
+        *> assert_eq (string.trim_left " ab ") "ab "
+        *> assert_eq (string.trim_right " ab ") " ab"
 
 let from_utf8_tests =
-    assert_req (string.from_utf8 []) (Ok "") *> assert_req (string.from_utf8 [32b]) (Ok " ")
-        *> assert_req (string.from_utf8 [195b, 165b, 195b, 164b, 195b, 182b]) (Ok "åäö")
-        *> assert_req (string.from_utf8 [195b, 165b, 195b, 164b, 195b]) (Err ())
-        *> assert_req (string.from_utf8 [195b, 165b, 195b, 195b, 182b]) (Err ())
+    assert_eq (string.from_utf8 []) (Ok "") *> assert_eq (string.from_utf8 [32b]) (Ok " ")
+        *> assert_eq (string.from_utf8 [195b, 165b, 195b, 164b, 195b, 182b]) (Ok "åäö")
+        *> assert_eq (string.from_utf8 [195b, 165b, 195b, 164b, 195b]) (Err ())
+        *> assert_eq (string.from_utf8 [195b, 165b, 195b, 195b, 182b]) (Err ())
 
 let tests =
-    slice_tests *> append_tests *> find_tests
+    slice_tests 
+        *> append_tests
+        *> find_tests
+        *> starts_ends_tests
+        *> trim_tests
```

Comment:
Oops, seems like some tests weren't run :)

Reply:
:) Need https://github.com/gluon-lang/gluon/issues/388 and then it would be possible to scan for unused variables.

Reply:
Auto test runner would help. Currently we need to do two steps: write a test case and don't forget to add it in `run`.

Reply:
Yeah. Perhaps the rudimentary attribute system I added with `@implicit` could be used with `@Test` attributes and procedural macros to rewrite the AST.

```f#
/// @Test
let test1 = ...

/// @Test
let test2 = ...

()
```
gets rewritten to
```f#
/// @Test
let test1 = ...

/// @Test
let test2 = ...
(test1 *> test2)
```

(if the `let`s don't end with `()` it could error.

Or perhaps a `test!` macro which takes a record.

```f#
test! {
    test1 = ...
    test2 = ...
}
```

```f#
test "test1 (..) *> test "test2" (..)
```

In any case it should still be easy to programmatically construct test. I want to further integrate the tests inside the `.glu` files with https://github.com/Marwes/tensile so that it is possible to see success and failure of each individual test case as well as filter on individual tests (currently it is only possible to filter on a per-file basis).

(Tensile is based on http://haskell.github.io/test-framework/)

Reply:
https://github.com/gluon-lang/gluon/issues/468